### ISSUE TITLE
Update docs with clarification on some important points

### ISSUE THREAD

**Issue by andrewhamon** [GUI/lua-resty-auto-ssl#100]
Hey there, great library! I'm really excited to use it. There's a few questions I have though that I think are worth clarifying in the readme, mostly relating to production and scale:

#### Does the `allow domain` hook run on every request?
Or does it only run if no certificate can be found? If it runs on every request, its important that it is fast.

#### Is the shared memory block used as a cache?
Or must all certificates be able to fit into memory? What happens when memory fills up? I'm hoping its used as a cache and a much larger set of certificates can be used than can fit into memory at once. This is important when you have, eg, multiple servers all using redis shared storage.

#### Is there a locking mechanism on certificate renewal?
This should be part of the storage adapter, in my opinion. If multiple servers are using a shared storage provider, there should ideally be a locking mechanism in place to prevent multiple servers from renewing a certificate at the same time.

**Comment by ecaroth** [comment]
I am also curious about the 3rd point on here, as I'm setting up an nginx reverse proxy with horizontal scaling using a shared volume for cert/lockfile sharing. Ideally *multiple* incoming requests to a hostname will block if _one_ request kicks off the cert request until the cert is available. Could run into LetsEncrypt throttling issues if multiple requests kick off the cert request during the verification process, and some initial requests failing.

**Comment by brianlund** [comment]
> Does the allow domain hook run on every request?
> Or does it only run if no certificate can be found? If it runs on every request, its important that it is fast.

It runs every time you establish a SSL/TLS session, not on every request. You can see this yourself if you add some logging to the function and call up a website. It's probably still a good idea to keep the function fast.

**Comment by luto** [comment]
> I've added a check for existing certificates as the first step in allow_domain and only if none is found does it check if the domain is allowed.

Would you mind sharing that, please? That's a nice example for the readme.

**Comment by brianlund** [comment]
> Would you mind sharing that, please? That's a nice example for the readme.
Certainly, I've shared it here before: https://github.com/GUI/lua-resty-auto-ssl/issues/94

```
-- first check if we have the certificate already, else check with backend if we allow it 
    local certstorage = auto_ssl:get("storage")

    local fullchain_pem, privkey_pem = certstorage:get_cert(domain)
    if fullchain_pem then
      return true
    else ...
```

**Comment by kapouer** [comment]
It seems it has been fixed in https://github.com/GUI/lua-resty-auto-ssl/commit/3515bb173c295278e8f7fd09ad0826def88fca97 (released in version 0.12.0). Isn't it ?

**Comment by brianlund** [comment]
> It seems it has been fixed in 3515bb1 (released in version 0.12.0). Isn't it ?

I agree

As for the 3th question: _is renewal locked across multiple servers?_

It is locked locally across multiple workers and on the storage adapter, which should lock it across multiple servers:

[/lib/resty/auto-ssl/jobs/renewal.lua](../blob/master/lib/resty/auto-ssl/jobs/renewal.lua)

```
local function renew_check_cert(auto_ssl_instance, storage, domain)
  -- Before issuing a cert, create a local lock to ensure multiple workers
  -- don't simultaneously try to register the same cert.
  ...
  -- Also add a lock to the configured storage adapter, which allows for a
  -- distributed lock across multiple servers (depending on the storage
  -- adapter).
```

**Comment by brendon** [comment]
Just a quick question RE the locking for certification creation. I had a look at the logs here:

https://crt.sh/?q=%25.spikeatschool.co.nz

It looks like the server requests two new certs per domain at the same time. Or am I mis-reading this page?

**Comment by andrewhamon** [comment]
@brendon looking at crt.sh for some of my domains, if you click on the id to get more details it looks like there is a `Summary: Precertificate` and `Summary: Leaf certificate`. No idea what those mean, but seems like something other than duplicate issuance :)
